Functor documentation and a conversion utility

GiudGiud opened the pull request.

GiudGiud committed:
Break out functor materials from materials

GiudGiud committed:
Break out functor materials from materials in modules

GiudGiud committed:
Use a new action for functor materials

GiudGiud committed:
Use a new FunctorMaterials syntax for functor materials

GiudGiud committed:
Move and add a little documentation to functor materials

GiudGiud committed:
Add Functor documentation
refs #24419

GiudGiud committed:
Add converter to go from functors to regular material properties
refs #19420

GiudGiud committed:
Add test for converting functor to regular material properties
refs #19420

GiudGiud committed:
Add exception checking
Add test for converting vector functors
refs #19420

lindsayad commented on framework/doc/content/source/materials/MaterialFunctorConverter.md:
```diff
@@ -0,0 +1,10 @@
+# MaterialFunctorConverter
+
+The `MaterialFunctorConverter` is used to explicitly convert functors, such as [functor material properties](syntax/FunctorMaterials/index.md)
+into regular nonAD or AD material properties.
```
```suggestion
into regular non-AD or AD material properties.
```

lindsayad commented on framework/doc/content/syntax/Materials/index.md:
```diff
@@ -222,142 +222,8 @@ error.
-other spatial argument types, evaluation of the functor (material property) may be relatively
-inexpensive compared to the memory expense incurred from caching. We may definitely implement
-caching for other overloads, however, if use cases call for it.
+Functor materials are special kind of materials used for on-the-fly material property evaluation.
```
```suggestion
Functor materials are a special kind of materials used for on-the-fly material property evaluation.
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+# FunctorMaterials
+
+Functor material properties are properties that are evaluated on-the-fly.
+Along with numerous classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
```
```suggestion
Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
```
Only three main systems are functors at the current time, so numerous feels like a bit too much. Maybe not forever though!

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+functor materials, etc, may consume.
+Both regular and functor material properties support a wide array of data types (scalar, vector, tensors,
+[automatic differentiation](automatic_differentiation/index.md) versions of all of these). Functor materials, because they are evaluated on-the-fly
+and call other functors on-the-fly as well, do not require dependency ordering.
```
Yes, another very important benefit which I often forget to bring up. Thanks for highlighting this!

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+[MaterialFunctorConverter.md] to process them.
+
+Functor materials can be created within the `[FunctorMaterials]` block, and currently the `[Materials]` block as well,
+but that syntax is considered for deprecation.
```
```suggestion
but that syntax will be deprecated.
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+on-the-fly. E.g. they can be viewed as functions of the current location in
+space (and time). Functor material properties provide several overloads of the
+`operator()` method for different "geometric quantities". One example of a
+"geometric quantity" is a `const Elem *`, e.g. for an `FVElementalKernel`, the
```
```suggestion
"geometric quantity" is based around an element, e.g. for an `FVElementalKernel`, the
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+value of a functor material property in a cell-averaged sense can be obtained by
+the syntax
+
+- `_foo(_current_elem)`
```
```suggestion
- `_foo(makeElemArg(_current_elem), determineState())`
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+
+- `_foo(_current_elem)`
+
+where here `_foo` is a functor material property data member of the kernel. The
```
```suggestion
where here `_foo` is a functor data member of the kernel, `makeElemArg` is a helper routine for creating a functor element-based spatial argument, and `determineState()` is a helper routine for determining the correct time state to evaluate at, e.g. the current time for an implicit kernel and the old time for an explicit kernel. The
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+Currently, functor material property evaluations are defined using the API:
+
+```c++
+template <typename T>
+template <typename PolymorphicLambda>
+void FunctorMaterialProperty<T>::
+setFunctor(const MooseMesh & mesh,
+           const std::set<SubdomainID> & block_ids,
+           PolymorphicLambda my_lammy);
````
```suggestion
template <typename T, typename PolymorphicLambda>
const Moose::FunctorBase<T> &
FunctorMaterial::addFunctorProperty(const std::string & name,
                                    PolymorphicLambda my_lammy,
                                    const std::set<ExecFlagType> & clearance_schedule = {
                                        EXEC_ALWAYS});
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+           PolymorphicLambda my_lammy);
+```
+
+where the first two arguments are used to setup block restriction and the last argument is a lambda
+defining the property evaluation. The lambda must be callable with two arguments, the first
````
```suggestion
where the first argument will be the functor name stored in the problem functor database (all functor names must be unique), the second argument is a lambda
defining the property evaluation, and the third optional argument is a set defining at which execution stages the functor evaluation cache should be cleared. The lambda must be callable with two arguments, the first
```

lindsayad commented:
Can you let me know when you're done pushing commits/when this is ready for review? I was in the middle of reviewing and then I was told I needed to refresh because of your force push.

GiudGiud commented:
oh force push was just for the design file in the tests. No change anywhere else
my b

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+an `ADReal` looks like:
+
+```c++
+    _constant_unity_prop.setFunctor(
````
```suggestion
    addFunctorProperty<Real>(
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+defining the property evaluation. The lambda must be callable with two arguments, the first
+corresponding to space, and the second corresponding to time, and must return the type `T` of the
+`FunctorMaterialProperty`. An example of setting a constant functor material property that returns
+an `ADReal` looks like:
```
```suggestion
a `Real` looks like:
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+
+```c++
+    _constant_unity_prop.setFunctor(
+        _mesh, blockIDs(), [](const auto &, const auto &) -> ADReal { return 1.; });
````
```suggestion
        "foo", [](const auto &, const auto &) -> Real { return 1.; });
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+An example of a functor material property that depends on a nonlinear variable would look like
+
+```c++
+    _u_prop.setFunctor(_mesh, blockIDs(), [this](const auto & r, const auto & t) -> ADReal {
+      return _u_var(r, t);
+    });
````
```suggestion
  addFunctorProperty<ADReal>(_density_name,
                             [this](const auto & r, const auto & t) -> ADReal
                             { return _fluid.rho_from_p_T(_pressure(r, t), _temperature(r, t)); });
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+    });
+```
+
+In the above example, we simply forward the calling arguments along to the variable. Variable
````
```suggestion
In the above example, we simply forward the calling arguments along to the captured functors. Variable
```

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,169 @@
+`_foo(const SpatialArg & r, const TemporalArg & t)`. Below are the possible type overloads of
+`SpatialArg`.
+
+### FaceArg id=spatial-overloads
```
Maybe we should move all of this to the new Functors `index.md`. The following will also need to be updated

lindsayad replied:
And I'm happy to help with the updates

lindsayad replied:
Oh it looks like we already do have it there. So let's remove this duplicate stuff

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
````diff
@@ -0,0 +1,169 @@
+        _mesh, blockIDs(), [](const auto &, const auto &) -> ADReal { return 1.; });
+```
+
+An example of a functor material property that depends on a nonlinear variable would look like
````
```suggestion
An example of a functor material property that depends on a fluid properties user object and pressure and temperature functors looks like
```

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,94 @@
+# Functor system
+
+Functors are an abstraction, existing as a base class, that is available to numerous systems in MOOSE:
```
```suggestion
Functors are an abstraction, existing as a base class, that is available to several systems in MOOSE:
```

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,94 @@
+`_foo(const SpatialArg & r, const TemporalArg & t)`. Below are the possible type overloads of
+`SpatialArg`.
+
+### FaceArg id=spatial-overloads
```
Need to update all of these. I can make a pull request against your branch if you want some help sharing the load here

GiudGiud replied:
can do it

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,94 @@
+By default, functors are always (re-)evaluated every time they are called with
+`operator()`. However, the base class of functors, `Moose::Functor`, has a
+`setCacheClearanceSchedule(const std::set<ExecFlagType> & clearance_schedule)` API that allows
+control of evaluations. Supported values for the `clearance_schedule` are any combination of
```
```suggestion
control of evaluations (in addition to the `clearance_schedule` argument to `addFunctorProperty` introduced above). Supported values for the `clearance_schedule` are any combination of
```

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,94 @@
+## Functor caching id=caching
+
+By default, functors are always (re-)evaluated every time they are called with
+`operator()`. However, the base class of functors, `Moose::Functor`, has a
```
```suggestion
`operator()`. However, the base class of functors, `Moose::FunctorBase`, has a
```

lindsayad commented on framework/include/materials/MaterialFunctorConverter.h:
```diff
@@ -0,0 +1,41 @@
+  /// Number of material properties to create
+  const std::size_t _num_functors_to_convert;
+
+  /// Incoming functors to convert
```
```suggestion
  /// Incoming functors to convert. We up-convert non-AD functors to AD functors to ease implementation
```

lindsayad commented on framework/src/materials/MaterialFunctorConverter.C:
```diff
@@ -0,0 +1,91 @@
+MaterialFunctorConverterTempl<T>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Converts functor to nonAD and AD regular material properties");
```
```suggestion
  params.addClassDescription("Converts functor to non-AD and AD regular material properties");
```

lindsayad commented on framework/src/materials/MaterialFunctorConverter.C:
```diff
@@ -0,0 +1,91 @@
+  : Material(parameters),
+    _num_functors_to_convert(getParam<std::vector<MooseFunctorName>>("functors_in").size())
+{
+  auto functors_in = getParam<std::vector<MooseFunctorName>>("functors_in");
+  auto reg_props_out = getParam<std::vector<MaterialPropertyName>>("reg_props_out");
+  auto ad_props_out = getParam<std::vector<MaterialPropertyName>>("ad_props_out");
```
```suggestion
  const auto & functors_in = getParam<std::vector<MooseFunctorName>>("functors_in");
  const auto & reg_props_out = getParam<std::vector<MaterialPropertyName>>("reg_props_out");
  const auto & ad_props_out = getParam<std::vector<MaterialPropertyName>>("ad_props_out");
```

lindsayad commented on framework/src/materials/MaterialFunctorConverter.C:
```diff
@@ -0,0 +1,91 @@
+void
+MaterialFunctorConverterTempl<T>::computeQpProperties()
+{
+  // TODO: do we know when we are on a face or an element?
+  const Moose::ElemQpArg arg = {_current_elem, _qp, _qrule};
```
We should check `_bnd`. If `true`, then we need to make an `ElemSideQpArg` instead

GiudGiud replied:
done

lindsayad commented on test/tests/materials/functor_conversion/tests:
```diff
@@ -0,0 +1,40 @@
+      input = conversion.i
+      cli_args = "Materials/convert_to_reg/reg_props_out='m1 m2 m3'"
+      expect_err = 'The number of output properties must match the number of input functors'
+      detail = 'if the user inputs a different number of functors and material properties, as the conversion match one to one, and'
```
I expect another detail with this terminating `and`

GiudGiud replied:
```suggestion
      detail = 'if the user inputs a different number of functors and material properties, as the conversion match one to one.'
```

lindsayad commented:
Ok done with this round of review!

GiudGiud committed:
Address Alex's review

- update all the functor coding explanations in docs

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1554727/) on e7cd336 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24494/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1554768/) on e7cd336 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24494 e7cd33</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.36%</td>
      <td>85.38%</td>
      <td>+0.01%</td>
      <td>96.54%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88359</td>
      <td>88415</td>
      <td>+56</td>
      <td>251</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15150</td>
      <td>15145</td>
      <td>-5</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24494/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24494/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24494 e7cd33</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.85%</td>
      <td>83.85%</td>
      <td>-</td>
      <td>99.15%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10785</td>
      <td>10785</td>
      <td>-</td>
      <td>234</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2078</td>
      <td>2078</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24494/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24494/coverage/navier_stokes)

### Solid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24494 e7cd33</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.87%</td>
      <td>83.87%</td>
      <td>-</td>
      <td>95.65%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>286</td>
      <td>286</td>
      <td>-</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>55</td>
      <td>55</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24494/coverage_diff/solid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24494/coverage/solid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24494/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Add support for boundary and neighbor material property evaluations
Use Qp 0 to simulate constant per element / subdomain behavior

GiudGiud committed:
Add test for face material properties converted from functors

GiudGiud committed:
Add rest of functor arguments to user docs
Rework sectioning
Add back Alex's suggestions after git mishap

GiudGiud commented:
Should be better now.
My apologies for not reading through the massive copy paste of outdated docs

GiudGiud committed:
Address doc failures:
- changed links
Add link to Functor index page in Functions and variables
Add object list in Positions

joshuahansel commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,138 @@
+value of a functor material property in a cell-averaged sense can be obtained by
+the syntax
+
+- `_foo(makeElemArg(_current_elem), determineState())`
```
Before this, you should talk about declaring the member variable (and what type).

GiudGiud replied:
You're right. I need to move more docs from FunctorMaterials.

joshuahansel reviewed (changes requested):
I still need to review more - I just wanted to take a break. I'm just reviewing the doc by the way.

lindsayad commented on framework/doc/content/syntax/Functions/index.md:
```diff
@@ -37,6 +37,20 @@ Check out `PiecewiseBilinear` to see how to update a function to support AD by
+of objects that can use `Functions` to compute spatial quantities.
+
+!alert note
+When making a new object using `Functions` to contribute back to MOOSE,
+we ask that you consider using [Functors](syntax/Functors/index.md) instead
+to naturally enable its use with variables and functor material properties.
```
Very nice note

joshuahansel replied:
Yeah, it's mainly something we need to communicate to reviewers.

lindsayad commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+has different methods to declare/get non-AD and AD properties, the new functor
+system has single APIs for both non-AD and AD property types.
+
+Currently, functor material property evaluations are defined using the API:
```
It hasn't changed for a while (although it did change once before), so probably not

GiudGiud replied:
Removed current

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,138 @@
+
+## Developing with functors
+
+Functors are evaluated on-the-fly. E.g. they can be viewed as functions of the current location in
+space (and time). Functors provide several overloads of the
+`operator()` method for different "geometric quantities". One example of a
+"geometric quantity" is based around an element, e.g. for an `FVElementalKernel`, the
+value of a functor material property in a cell-averaged sense can be obtained by
+the syntax
+
+- `_foo(makeElemArg(_current_elem), determineState())`
+
+where here `_foo` is a functor data member of the kernel, `makeElemArg` is a helper routine for creating a
+functor element-based spatial argument, and `determineState()` is a helper routine for determining the correct
+time state to evaluate at, e.g. the current time for an implicit kernel and the old time for an explicit kernel.
```
I believe this is all in the `FunctorMaterials/index.md` page as well. I think some duplication is fine, although this seems like a lot

GiudGiud replied:
I ll go back and check

GiudGiud replied:
made a link and more subsections

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,138 @@
+
+#### FaceArg
+
+A typedef defining a "face" evaluation calling argument. This is composed of
```
```suggestion
A `struct` defining a "face" evaluation calling argument. This is composed of
```
If we want to be C++ technical

lindsayad commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,138 @@
+  interpolated to the face
+- a boolean which states whether the face information element is upwind of the face
+- a boolean to indicate whether to correct for element skewness
+- a pair of subdomain IDs. These do not always correspond to the face info element subdomain
+  ID and face info neighbor subdomain ID. For instance if a flux kernel is operating at a
+  subdomain boundary on which the kernel is defined on one side but not the other, the
+  passed-in subdomain IDs will both correspond to the subdomain ID that the flux kernel is
+  defined on
```
This has been replaced with a `const Elem * face_side` data member

GiudGiud replied:
changed

lindsayad commented on framework/src/materials/MaterialFunctorConverter.C:
```diff
@@ -0,0 +1,106 @@
+
+  if (isParamValid("reg_props_out") && isParamValid("ad_props_out"))
+    paramError("reg_props_out",
+               "We dont support converting functors to both AD and regular material properties. "
+               "Please use another " +
+                   type() + " to convert to both types.");
```
```suggestion
               "We dont support converting functors to both regular and AD material properties in a single instance of '", type(), "'. Please create two instances, one for regular and one for AD.");
```

GiudGiud replied:
done

GiudGiud commented on framework/doc/content/syntax/Functions/index.md:
```diff
@@ -37,6 +37,20 @@ Check out `PiecewiseBilinear` to see how to update a function to support AD by
+objects that can compute values at a location in space and time.
+
+As `Functors`, they may be specified to objects such as the
+[FunctorElementalAux.md] in their functor argument. This vastly expands the number
```
```suggestion
[FunctorElementalAux.md] in their [!param](/AuxKernels/FunctorElementalAux/functor) parameter. This vastly expands the number
```

GiudGiud commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+# FunctorMaterials
+
+Functor material properties are properties that are evaluated on-the-fly.
+Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
+for the creation of objects with considerable inter-operability between systems.
+
+## Using functor materials
+
+Much like regular materials declare regular material properties, functor materials declare functor material properties.
```
```suggestion

Much like regular materials declare regular material properties, functor materials declare functor material properties.
Functor material properties are properties that are evaluated on-the-fly.
Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
for the creation of objects with considerable inter-operability between systems.

## Using functor materials

```

GiudGiud commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+
+!alert note
+If a Functor is reported as missing by the simulation, and it was supposed to be created by a `FunctorMaterial`,
+you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were
```
```suggestion
you may use the [!param](/Debug/SetupDebugAction/show_functors) parameter to get more information about what functors were
```

joshuahansel replied:
What I meant was that there's no indication that this is a parameter that needs to be in a `Debug` block.

GiudGiud commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+has different methods to declare/get non-AD and AD properties, the new functor
+system has single APIs for both non-AD and AD property types.
+
+Currently, functor material property evaluations are defined using the API:
```
```suggestion
Functor material property evaluations are defined using the API:
```

GiudGiud commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -0,0 +1,138 @@
+# Functor system
+
+Functors are an abstraction, existing as a base class, that is available to several systems in MOOSE:
```
It's more that they can be used as functors.
It s more an added capability than a restriction. Maybe there's a better word to express that

GiudGiud committed:
Fixup documentation more
- add docs on header + constructor for class using functors
Co-authored-by: joshuahansel <joshua.hansel@inl.gov>
Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+
+!alert note
+If a Functor is reported as missing by the simulation, and it was supposed to be created by a `FunctorMaterial`,
+you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were
```
added parameter after

joshuahansel replied:
I put this comment on some outdated suggestion I think: the problem is more that there's no indication that this "show_functors" parameter is a `Debug` block parameter.

GiudGiud replied:
will change it

GiudGiud commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+
+## Using functor materials
+
+Much like regular materials declare regular material properties, functor materials declare functor material properties.
```
done

GiudGiud commented on framework/doc/content/syntax/FunctorMaterials/index.md:
```diff
@@ -0,0 +1,108 @@
+disabled by default due to the potential memory cost. The reader is referred to the [Functors caching documentation](syntax/Functors/index.md#caching)
+
+!alert warning
+Functor materials are generally **NOT** compatible with regular materials, as regular material properties **CANNOT** be used
```
reworded

lindsayad reviewed (dismissed):
great stuff

joshuahansel commented:
Is the conversion utility news-worthy?

GiudGiud committed:
Address Josh's review:
- clear indication of parameter belonging
- fix syntax links

GiudGiud commented:
I ll add it to Casey's PR directly. I will hopefully get two more PRs in today so I can do them all at once

GiudGiud commented:
please let me know when you're good with this review, I ll push all the changes together

joshuahansel commented:
> please let me know when you're good with this review, I ll push all the changes together

I'll have more changes that I'd like, but nothing here is inaccurate - just potential improvements, so I'll approve after you implement the requested changes, and then at some point I'll probably make my own functor doc improvement PR after I get my feet wet with them and understand them more.

GiudGiud commented:
Improved documentation after your experience sounds ideal!
